In [101]:
#Task 1.1 Variable Roles and Data Types
import pandas as pd
import numpy as np

df = pd.read_csv('dataset_ifn580_262_v1.csv')

columns_to_check = [
    'destination',
    'passanger',
    'weather',
    'coupon',
    'expiration',
    'gender',
    'maritalStatus',
    'education',
    'occupation'
]

for col in columns_to_check:
    print(col, ":", df[col].unique())

destination : <StringArray>
['No Urgent Place', 'Home', 'Work']
Length: 3, dtype: str
passanger : <StringArray>
['Alone', 'Friend(s)', 'Kid(s)', 'Partner']
Length: 4, dtype: str
weather : <StringArray>
['Sunny', 'Rainy', 'Snowy']
Length: 3, dtype: str
coupon : <StringArray>
[      'Restaurant(<20)',          'Coffee House', 'Carry out & Take away',
                   'Bar',     'Restaurant(20-50)']
Length: 5, dtype: str
expiration : <StringArray>
['1d', '2h']
Length: 2, dtype: str
gender : <StringArray>
['Female', 'Male']
Length: 2, dtype: str
maritalStatus : <StringArray>
['Unmarried partner', 'Single', 'Married partner', 'Divorced', 'Widowed']
Length: 5, dtype: str
education : <StringArray>
[              'Some college - no degree',
                       'Bachelors degree',
                      'Associates degree',
                   'High School Graduate',
 'Graduate degree (Masters or Doctorate)',
                       'Some High School']
Length: 6, dtype: str
occupation : <Stri

# Missing Value

- 'car' has approximately 99.15% missing values and may be unreliable for modelling.
- 'temperature' has a small proportion of missing values.
- 'Bar', 'CoffeeHouse', 'CarryAway', 'RestaurantLessThan20', and 'Restaurant20To50' also have small proportions of missing values.
- The restaurant/visit-frequency variables contain ordered categories, so they should be treated as ordinal rather than nominal.

'Temperture' float64/ 1.970987 [55. 80. nan 30.]

'Bar' object/ 0.843582 

'Coffeehouse' object/ 1.710817

'Carryaway' object/ 1.190476

'RestaurantLessThan20' object/ 1.024913

'Restaurant20To50' object/ 1.490066

# Data exploration

Age:
[21, 46, 26, 31, 41, 50plus, 36, below21]
- object in pandas
- contains numeric categories plus textual boundaries such as below21 / 50plus
- ordered meaning
- likely ordinal

income:
[$37500-$49999, $62500-$74999, $12500-$24999, $75000-$87499, $50000-$62499, $25000-$37499, $100000 or more, $87500-$99999, less than $12500]
- object in pandas
- income range represented as text
- ordered meaning
- likely ordinal

time:
[2PM, 10AM, 6PM, 7AM, 10PM]
- object in pandas
- clock time represented as text with AM/PM
- requires numerical/ordered representation before modelling

In [102]:
# Task 1.2 Data Quality, Cleaning and Preprocessing
missing_count = df.isnull().sum()
missing_percent = df.isnull().mean() * 100

missing_summary = pd.DataFrame({
    'Missing Count': missing_count,
    'Missing Percentage': missing_percent
})

missing_summary = missing_summary[missing_summary['Missing Count'] > 0]

display(missing_summary)

,Missing Count,Missing Percentage
temperature,250,1.970987
car,12576,99.148534
Bar,107,0.843582
CoffeeHouse,217,1.710817
CarryAway,151,1.190476
RestaurantLessThan20,130,1.024913
Restaurant20To50,189,1.490066


Car:
- missing rate> 99%
- too little observed information
- Impuation would be unreliable
- Remove the variable

In [103]:
print("Mean:", df['temperature'].mean())
print("Median:", df['temperature'].median())

Mean: 63.28976998552356
Median: 80.0


Temperature:
- has 250 missing values, which is a relatively small proportion of the dataset.
- The mean is 63.289, while the median is 80.
- The original temperature values are 30, 55, and 80.
- Median imputation is selected because it keeps the imputed value within the existing temperature values and avoids introducing a new value that does not appear in the original dataset.

In [104]:
print(df[col].value_counts(dropna=False))

occupation
Unemployed                                   1870
Student                                      1584
Computer & Mathematical                      1408
Sales & Related                              1093
Education&Training&Library                    943
Management                                    838
Office & Administrative Support               639
Arts Design Entertainment Sports & Media      629
Business & Financial                          544
Retired                                       495
Food Preparation & Serving Related            298
Healthcare Practitioners & Technical          244
Healthcare Support                            242
Community & Social Services                   241
Legal                                         219
Transportation & Material Moving              218
Architecture & Engineering                    175
Personal Care & Service                       175
Protective Service                            175
Life Physical Social Science           

Bar:
- has 107 missing values, which is a small proportion of the dataset.
- The variable contains ordered frequency categories such as never, less1, 1~3, 4~8, and gt8.
- The most frequent category is never
- Mode imputation is selected because the variable is categorical and the missing proportion is low.

In [105]:
print(df[col].value_counts(dropna=False))

occupation
Unemployed                                   1870
Student                                      1584
Computer & Mathematical                      1408
Sales & Related                              1093
Education&Training&Library                    943
Management                                    838
Office & Administrative Support               639
Arts Design Entertainment Sports & Media      629
Business & Financial                          544
Retired                                       495
Food Preparation & Serving Related            298
Healthcare Practitioners & Technical          244
Healthcare Support                            242
Community & Social Services                   241
Legal                                         219
Transportation & Material Moving              218
Architecture & Engineering                    175
Personal Care & Service                       175
Protective Service                            175
Life Physical Social Science           

CoffeeHouse:
- has 217 missing values, which is a small proportion of the dataset.
- The variable contains ordered frequency categories such as never, less1, 1~3, 4~8, and gt8.
- The most frequent category is less1
- Mode imputation is selected because the variable is categorical and the missing proportion is low.

In [106]:
print(df[col].value_counts(dropna=False))

occupation
Unemployed                                   1870
Student                                      1584
Computer & Mathematical                      1408
Sales & Related                              1093
Education&Training&Library                    943
Management                                    838
Office & Administrative Support               639
Arts Design Entertainment Sports & Media      629
Business & Financial                          544
Retired                                       495
Food Preparation & Serving Related            298
Healthcare Practitioners & Technical          244
Healthcare Support                            242
Community & Social Services                   241
Legal                                         219
Transportation & Material Moving              218
Architecture & Engineering                    175
Personal Care & Service                       175
Protective Service                            175
Life Physical Social Science           

CarryAway:
- has 151 missing values, which is a small proportion of the dataset.
- The variable contains ordered frequency categories such as never, less1, 1~3, 4~8, and gt8.
- The most frequent category is 1~3.
- Mode imputation is selected because the variable is categorical and the missing proportion is low.

In [107]:
print(df[col].value_counts(dropna=False))

occupation
Unemployed                                   1870
Student                                      1584
Computer & Mathematical                      1408
Sales & Related                              1093
Education&Training&Library                    943
Management                                    838
Office & Administrative Support               639
Arts Design Entertainment Sports & Media      629
Business & Financial                          544
Retired                                       495
Food Preparation & Serving Related            298
Healthcare Practitioners & Technical          244
Healthcare Support                            242
Community & Social Services                   241
Legal                                         219
Transportation & Material Moving              218
Architecture & Engineering                    175
Personal Care & Service                       175
Protective Service                            175
Life Physical Social Science           

RestaurantLessThan20:
- has 130 missing values, which is a small proportion of the dataset.
- The variable contains ordered frequency categories such as never, less1, 1~3, 4~8, and gt8.
- The most frequent category is 1~3.
- Mode imputation is selected because the variable is categorical and the missing proportion is low.

In [108]:
print(df[col].value_counts(dropna=False))

occupation
Unemployed                                   1870
Student                                      1584
Computer & Mathematical                      1408
Sales & Related                              1093
Education&Training&Library                    943
Management                                    838
Office & Administrative Support               639
Arts Design Entertainment Sports & Media      629
Business & Financial                          544
Retired                                       495
Food Preparation & Serving Related            298
Healthcare Practitioners & Technical          244
Healthcare Support                            242
Community & Social Services                   241
Legal                                         219
Transportation & Material Moving              218
Architecture & Engineering                    175
Personal Care & Service                       175
Protective Service                            175
Life Physical Social Science           

Restaurant20To50:
- has 189 missing values, which is a small proportion of the dataset.
- The variable contains ordered frequency categories such as never, less1, 1~3, 4~8, and gt8.
- The most frequent category is less1.
- Mode imputation is selected because the variable is categorical and the missing proportion is low.

In [109]:
# Task 1.3 Mixed Numerical/Textual Representation and Time

df = df.drop(columns=['car'])
df['temperature'] = df['temperature'].fillna(df['temperature'].median())
df['Bar'] = df['Bar'].fillna(df['Bar'].mode()[0])
df['CoffeeHouse'] = df['CoffeeHouse'].fillna(df['CoffeeHouse'].mode()[0])
df['CarryAway'] = df['CarryAway'].fillna(df['CarryAway'].mode()[0])
df['RestaurantLessThan20'] = df['RestaurantLessThan20'].fillna(df['RestaurantLessThan20'].mode()[0])
df['Restaurant20To50'] = df['Restaurant20To50'].fillna(df['Restaurant20To50'].mode()[0])

df.isnull().sum()

destination             0
passanger               0
weather                 0
temperature             0
time                    0
coupon                  0
expiration              0
gender                  0
age                     0
maritalStatus           0
has_children            0
education               0
occupation              0
income                  0
Bar                     0
CoffeeHouse             0
CarryAway               0
RestaurantLessThan20    0
Restaurant20To50        0
toCoupon_GEQ5min        0
toCoupon_GEQ15min       0
toCoupon_GEQ25min       0
direction_same          0
direction_opp           0
is_accepted             0
dtype: int64

In [110]:
# encoding(maping)
age_mapping = {
    'below21': 0,
    '21':1,
    '26':2,
    '31':3,
    '36':4,
    '41':5,
    '46':6,
    '50plus': 7
}
df['age'] = df['age'].map(age_mapping)

In [111]:
df['age'].unique()

array([1, 6, 2, 3, 5, 7, 4, 0])

In [112]:
df['age'].isnull().sum()

np.int64(0)

In [113]:
time_mapping = {
    '7AM': 0,
    '10AM': 1,
    '2PM': 2,
    '6PM': 3,
    '10PM': 4
}
df['time'] = df['time'].map(time_mapping)

### Time Encoding

- The `time` variable has a natural chronological order during the day.
- The categories were manually ordered from the earliest to the latest time.
- Consecutive integers were used to preserve this order in a numerical form.
- These encoded values represent ordinal positions only and do not represent the actual time intervals between categories.

In [114]:
df['time'].unique()

array([2, 1, 3, 0, 4])

In [115]:
df['time'].isnull().sum()

np.int64(0)

In [116]:
income_mapping = {
    'Less than $12500': 0,
    '$12500 - $24999': 1,
    '$25000 - $37499': 2,
    '$37500 - $49999': 3,
    '$50000 - $62499': 4,
    '$62500 - $74999': 5,
    '$75000 - $87499': 6,
    '$87500 - $99999': 7,
    '$100000 or More': 8
}
df['income'] = df['income'].map(income_mapping)

In [117]:
df['income'].unique()

array([3, 5, 1, 6, 4, 2, 8, 7, 0])

In [118]:
df['income'].isnull().sum()

np.int64(0)

In [119]:
df['education'].unique()

<StringArray>
[              'Some college - no degree',
                       'Bachelors degree',
                      'Associates degree',
                   'High School Graduate',
 'Graduate degree (Masters or Doctorate)',
                       'Some High School']
Length: 6, dtype: str

In [120]:
education_mapping = {
    'Some High School': 0,
    'High School Graduate': 1,
    'Some college - no degree': 2,
    'Associates degree': 3,
    'Bachelors degree': 4,
    'Graduate degree (Masters or Doctorate)': 5
}
df['education'] = df['education'].map(education_mapping)

In [121]:
df['education'].unique()

array([2, 4, 3, 1, 5, 0])

In [122]:
df['education'].isnull().sum()

np.int64(0)

In [123]:
frequency_mapping = {
    'never': 0,
    'less1': 1,
    '1~3': 2,
    '4~8': 3,
    'gt8': 4
}
frequency_columns = [
    'Bar',
    'CoffeeHouse',
    'CarryAway',
    'RestaurantLessThan20',
    'Restaurant20To50'
]

for col in frequency_columns:
    df[col] = df[col].map(frequency_mapping)

In [124]:
df[frequency_columns].isnull().sum()

Bar                     0
CoffeeHouse             0
CarryAway               0
RestaurantLessThan20    0
Restaurant20To50        0
dtype: int64

In [125]:
for col in frequency_columns:
    print(col, df[col].unique())

Bar [0 1 2 4 3]
CoffeeHouse [0 1 3 2 4]
CarryAway [2 3 4 1 0]
RestaurantLessThan20 [3 2 1 4 0]
Restaurant20To50 [2 1 0 4 3]


In [126]:
expiration_mapping = {
    '2h': 0,
    '1d': 1
}
df['expiration'] = df['expiration'].map(expiration_mapping)

In [127]:
df['expiration'].unique()

array([1, 0])

In [128]:
df['expiration'].isnull().sum()

np.int64(0)

In [129]:
gender_mapping = {
    'Female': 0,
    'Male': 1
}
df['gender'] = df['gender'].map(gender_mapping)

In [130]:
df['gender'].unique()

array([0, 1])

In [131]:
destination_encoded = pd.get_dummies(
    df['destination'],
    prefix='destination',
    dtype=int
)

### Nominal Variable Encoding

- `destination` is a nominal categorical variable with no natural order.
- One-hot encoding is used to convert the categories into numerical binary variables.
- Each category is represented by a separate 0/1 column.

In [132]:
destination_encoded.head()

,destination_Home,destination_No Urgent Place,destination_Work
0,0,1,0
1,0,1,0
2,0,1,0
3,0,1,0
4,0,1,0


In [133]:
df = pd.concat([df, destination_encoded], axis=1)

In [134]:
df = df.drop(columns=['destination'])

In [135]:
df.head()

,passanger,weather,temperature,time,coupon,expiration,gender,age,maritalStatus,has_children,...,Restaurant20To50,toCoupon_GEQ5min,toCoupon_GEQ15min,toCoupon_GEQ25min,direction_same,direction_opp,is_accepted,destination_Home,destination_No Urgent Place,destination_Work
0,Alone,Sunny,55.0,2,Restaurant(<20),1,0,1,Unmarried partner,1,...,2,1,0,0,0,1,1,0,1,0
1,Friend(s),Sunny,80.0,1,Coffee House,0,0,1,Unmarried partner,1,...,2,1,0,0,0,1,0,0,1,0
2,Friend(s),Sunny,80.0,1,Carry out & Take away,0,0,1,Unmarried partner,1,...,2,1,1,0,0,1,1,0,1,0
3,Friend(s),Sunny,80.0,2,Coffee House,0,0,1,Unmarried partner,1,...,2,1,1,0,0,1,0,0,1,0
4,Friend(s),Sunny,80.0,2,Coffee House,1,0,1,Unmarried partner,1,...,2,1,1,0,0,1,0,0,1,0


In [136]:
nominal_columns = [
    'passanger',
    'weather',
    'coupon',
    'maritalStatus',
    'occupation'
]
nominal_encoded = pd.get_dummies(
    df[nominal_columns],
    prefix=nominal_columns,
    dtype=int
)
df = pd.concat([df, nominal_encoded], axis=1)

In [137]:
df = df.drop(columns=nominal_columns)

In [138]:
df.head()

,temperature,time,expiration,gender,age,has_children,education,income,Bar,CoffeeHouse,...,occupation_Management,occupation_Office & Administrative Support,occupation_Personal Care & Service,occupation_Production Occupations,occupation_Protective Service,occupation_Retired,occupation_Sales & Related,occupation_Student,occupation_Transportation & Material Moving,occupation_Unemployed
0,55.0,2,1,0,1,1,2,3,0,0,...,0,0,0,0,0,0,0,0,0,1
1,80.0,1,0,0,1,1,2,3,0,0,...,0,0,0,0,0,0,0,0,0,1
2,80.0,1,0,0,1,1,2,3,0,0,...,0,0,0,0,0,0,0,0,0,1
3,80.0,2,0,0,1,1,2,3,0,0,...,0,0,0,0,0,0,0,0,0,1
4,80.0,2,1,0,1,1,2,3,0,0,...,0,0,0,0,0,0,0,0,0,1


In [139]:
df.dtypes

temperature                                    float64
time                                             int64
expiration                                       int64
gender                                           int64
age                                              int64
                                                ...   
occupation_Retired                               int64
occupation_Sales & Related                       int64
occupation_Student                               int64
occupation_Transportation & Material Moving      int64
occupation_Unemployed                            int64
Length: 64, dtype: object

# checking Object

- `df.select_dtypes(include='object').columns` was used to check whether any text-based columns remained after encoding.
- No `object` columns remain.
- `df.isnull().sum().sum()` was used to check the total number of missing values in the processed dataset.
- The result was 0, confirming that no missing values remain.

In [140]:
# Task 1.4 – Correlation Analysis and Redundant Variables

df.select_dtypes(exclude='number').columns

Index([], dtype='str')

In [141]:
df.dtypes[df.dtypes == 'object']

Series([], dtype: object)

In [142]:
correlation_matrix = df.corr()

In [143]:
correlation_matrix

,temperature,time,expiration,gender,age,has_children,education,income,Bar,CoffeeHouse,...,occupation_Management,occupation_Office & Administrative Support,occupation_Personal Care & Service,occupation_Production Occupations,occupation_Protective Service,occupation_Retired,occupation_Sales & Related,occupation_Student,occupation_Transportation & Material Moving,occupation_Unemployed
temperature,1.000000,-0.035530,-0.121833,-0.025419,-0.017893,-0.019894,0.008683,-0.029969,-0.005448,-0.016049,...,0.040296,-0.019941,-0.003830,-0.001028,0.022685,0.021477,-0.003243,0.022421,0.002254,-0.029577
time,-0.035530,1.000000,-0.030381,-0.003593,0.005114,0.026883,0.009751,0.010860,-0.014447,-0.000058,...,-0.020176,0.001148,0.003669,0.007005,-0.010955,-0.013038,0.000643,-0.008162,0.000166,0.019036
expiration,-0.121833,-0.030381,1.000000,0.001264,-0.021208,-0.016020,-0.016356,-0.020606,-0.009821,-0.002195,...,-0.014373,0.025240,-0.003858,0.017989,-0.007942,-0.007158,-0.002287,0.016554,0.000155,0.008767
gender,-0.025419,-0.003593,0.001264,1.000000,-0.099847,-0.159956,-0.004301,0.101738,0.161595,-0.044470,...,0.088371,-0.097346,-0.027272,0.058632,0.033578,0.013924,0.057932,0.092138,0.055712,-0.137959
age,-0.017893,0.005114,-0.021208,-0.099847,1.000000,0.442096,0.158249,0.084885,-0.158573,-0.049766,...,0.060981,0.025226,0.110139,0.067808,0.021135,0.313798,-0.000859,-0.396503,0.075858,-0.082651
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
occupation_Retired,0.021477,-0.013038,-0.007158,0.013924,0.313798,0.114056,0.051814,-0.002094,-0.132851,-0.085394,...,-0.053599,-0.046416,-0.023836,-0.018849,-0.023836,1.000000,-0.061883,-0.076126,-0.026649,-0.083800
occupation_Sales & Related,-0.003243,0.000643,-0.002287,0.057932,-0.000859,-0.046576,-0.032885,-0.063590,-0.009931,-0.050215,...,-0.081674,-0.070729,-0.036321,-0.028722,-0.036321,-0.061883,1.000000,-0.116002,-0.040608,-0.127696
occupation_Student,0.022421,-0.008162,0.016554,0.092138,-0.396503,-0.222232,-0.145635,-0.117955,-0.021115,0.125814,...,-0.100474,-0.087009,-0.044681,-0.035333,-0.044681,-0.076126,-0.116002,1.000000,-0.049955,-0.157088
occupation_Transportation & Material Moving,0.002254,0.000166,0.000155,0.055712,0.075858,0.048911,-0.066045,-0.002199,0.009791,-0.007444,...,-0.035172,-0.030459,-0.015641,-0.012369,-0.015641,-0.026649,-0.040608,-0.049955,1.000000,-0.054991


In [144]:
high_corr_pairs = []

for i in range(len(correlation_matrix.columns)):
    for j in range(i):
        corr_value = correlation_matrix.iloc[i, j]

        if abs(corr_value) > 0.90:
            high_corr_pairs.append((
                correlation_matrix.columns[i],
                correlation_matrix.columns[j],
                corr_value
            ))

high_corr_pairs

[('direction_opp', 'direction_same', np.float64(-1.0))]

# remove direction_opp

- direction_same and direction_opp are perfectly negatively correlated (r = -1.0).
- Both variables contain redundant information.
- direction_same is retained because it is more directly interpretable.
- direction_opp is removed.

In [145]:
df = df.drop(columns=['direction_opp'])

In [146]:
correlation_matrix = df.corr()

In [147]:
'direction_opp' in df.columns

False

# How to Use the Preprocessing Code(For teammate)

- Run the notebook sequentially from the first cell to the last cell.
- Do not repeatedly run the .map() encoding cells after the variables have already been encoded. Running the same mapping again on numerical values can convert them into NaN.
- If the Jupyter kernel is restarted or df is no longer defined, rerun the notebook from the cell that loads the original CSV file.
- Tasks 2, 3, and 4 should use the same final processed df from Task 1 so that all three models use a consistent base dataset.
- If a later model requires additional preprocessing, such as scaling for Logistic Regression or Neural Networks, this should be performed within the relevant task without changing the common Task 1 preprocessing procedure.
- The final Task 1 dataset has completed missing-value treatment, encoding, and correlation-based redundancy removal and can be used as the common preprocessed base dataset for Tasks 2–4.